File: lib/gym_pulse-stack.ts
import * as cdk from 'aws-cdk-lib';
import * as iot from 'aws-cdk-lib/aws-iot';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as apigatewayv2 from 'aws-cdk-lib/aws-apigatewayv2';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as location from 'aws-cdk-lib/aws-location';
import * as bedrock from 'aws-cdk-lib/aws-bedrock';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export class GymAvailabilityStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // DynamoDB Tables
    const machineStateTable = new dynamodb.Table(this, 'MachineStateTable', {
      tableName: 'gym-machine-states',
      partitionKey: { name: 'machineId', type: dynamodb.AttributeType.STRING },
      sortKey: { name: 'timestamp', type: dynamodb.AttributeType.NUMBER },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      timeToLiveAttribute: 'ttl',
      stream: dynamodb.StreamViewType.NEW_AND_OLD_IMAGES,
    });

    const aggregatesTable = new dynamodb.Table(this, 'AggregatesTable', {
      tableName: 'gym-aggregates',
      partitionKey: { name: 'pk', type: dynamodb.AttributeType.STRING }, // branch#category#date
      sortKey: { name: 'sk', type: dynamodb.AttributeType.STRING }, // 15min-bin
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
    });

    const alertsTable = new dynamodb.Table(this, 'AlertsTable', {
      tableName: 'gym-alerts',
      partitionKey: { name: 'machineId', type: dynamodb.AttributeType.STRING },
      sortKey: { name: 'userId', type: dynamodb.AttributeType.STRING },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
    });

    // IoT Core Setup
    const iotPolicy = new iot.CfnPolicy(this, 'GymDevicePolicy', {
      policyName: 'GymDevicePolicy',
      policyDocument: {
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Action: ['iot:Connect'],
            Resource: `arn:aws:iot:${this.region}:${this.account}:client/gym-*`,
          },
          {
            Effect: 'Allow',
            Action: ['iot:Publish'],
            Resource: `arn:aws:iot:${this.region}:${this.account}:topic/org/*/machines/*/status`,
          },
          {
            Effect: 'Allow',
            Action: ['iot:Subscribe'],
            Resource: `arn:aws:iot:${this.region}:${this.account}:topicfilter/org/*/machines/*/status`,
          },
        ],
      },
    });

    // Lambda Functions
    const ingestLambda = new lambda.Function(this, 'IngestLambda', {
      functionName: 'gym-ingest-handler',
      runtime: lambda.Runtime.PYTHON_3_10,
      handler: 'index.handler',
      code: lambda.Code.fromInline(`
import json
import boto3
import time
from decimal import Decimal

dynamodb = boto3.resource('dynamodb')
machine_table = dynamodb.Table('${machineStateTable.tableName}')
aggregates_table = dynamodb.Table('${aggregatesTable.tableName}')

def handler(event, context):
    try:
        # Parse IoT message
        topic = event['topic']
        payload = json.loads(event['payload'])
        
        machine_id = payload['machineId']
        status = payload['status']  # 'occupied' or 'free'
        timestamp = int(time.time())
        
        # Update current state
        machine_table.put_item(Item={
            'machineId': machine_id,
            'timestamp': timestamp,
            'status': status,
            'ttl': timestamp + 86400  # 24 hours TTL
        })
        
        # TODO: Update aggregates for heatmaps
        # TODO: Trigger alerts if status changed to 'free'
        
        return {'statusCode': 200}
    except Exception as e:
        print(f"Error: {str(e)}")
        return {'statusCode': 500}
      `),
      environment: {
        MACHINE_STATE_TABLE: machineStateTable.tableName,
        AGGREGATES_TABLE: aggregatesTable.tableName,
      },
      timeout: cdk.Duration.seconds(30),
    });

    const apiLambda = new lambda.Function(this, 'ApiLambda', {
      functionName: 'gym-api-handler',
      runtime: lambda.Runtime.PYTHON_3_10,
      handler: 'index.handler',
      code: lambda.Code.fromInline(`
import json
import boto3
from boto3.dynamodb.conditions import Key

dynamodb = boto3.resource('dynamodb')
machine_table = dynamodb.Table('${machineStateTable.tableName}')

def handler(event, context):
    try:
        path = event['path']
        method = event['httpMethod']
        
        if path == '/branches' and method == 'GET':
            # Return branch list with availability counts
            return {
                'statusCode': 200,
                'headers': {'Content-Type': 'application/json'},
                'body': json.dumps([
                    {
                        'id': 'branch-1',
                        'name': 'Central Branch',
                        'coordinates': [22.2783, 114.1747],
                        'categories': {
                            'legs': {'free': 3, 'total': 8},
                            'chest': {'free': 5, 'total': 10},
                            'back': {'free': 2, 'total': 6}
                        }
                    },
                    {
                        'id': 'branch-2', 
                        'name': 'Causeway Bay Branch',
                        'coordinates': [22.2783, 114.1830],
                        'categories': {
                            'legs': {'free': 1, 'total': 6},
                            'chest': {'free': 8, 'total': 12},
                            'back': {'free': 4, 'total': 8}
                        }
                    }
                ])
            }
        
        return {'statusCode': 404, 'body': 'Not found'}
    except Exception as e:
        return {'statusCode': 500, 'body': json.dumps({'error': str(e)})}
      `),
      environment: {
        MACHINE_STATE_TABLE: machineStateTable.tableName,
      },
      timeout: cdk.Duration.seconds(30),
    });

    // Tool functions for Bedrock Agent
    const availabilityToolLambda = new lambda.Function(this, 'AvailabilityToolLambda', {
      functionName: 'gym-availability-tool',
      runtime: lambda.Runtime.PYTHON_3_10,
      handler: 'index.handler',
      code: lambda.Code.fromInline(`
import json
import math

def haversine_distance(lat1, lon1, lat2, lon2):
    R = 6371  # Earth's radius in km
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = math.sin(dlat/2)**2 + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(dlon/2)**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1-a))
    return R * c

def handler(event, context):
    try:
        # Parse tool input
        body = json.loads(event['body']) if 'body' in event else event
        lat = float(body['lat'])
        lon = float(body['lon']) 
        radius = float(body.get('radius', 5))  # 5km default
        category = body['category']
        
        # Mock branches data (replace with DynamoDB query)
        branches = [
            {
                'id': 'branch-1',
                'name': 'Central Branch',
                'coordinates': [22.2783, 114.1747],
                'categories': {
                    'legs': {'free': 3, 'total': 8},
                    'chest': {'free': 5, 'total': 10},
                    'back': {'free': 2, 'total': 6}
                }
            },
            {
                'id': 'branch-2',
                'name': 'Causeway Bay Branch', 
                'coordinates': [22.2783, 114.1830],
                'categories': {
                    'legs': {'free': 1, 'total': 6},
                    'chest': {'free': 8, 'total': 12},
                    'back': {'free': 4, 'total': 8}
                }
            }
        ]
        
        # Filter by distance and category availability
        results = []
        for branch in branches:
            branch_lat, branch_lon = branch['coordinates']
            distance = haversine_distance(lat, lon, branch_lat, branch_lon)
            
            if distance <= radius and category in branch['categories']:
                results.append({
                    'branchId': branch['id'],
                    'name': branch['name'],
                    'coordinates': branch['coordinates'],
                    'distance': round(distance, 2),
                    'freeCount': branch['categories'][category]['free'],
                    'totalCount': branch['categories'][category]['total']
                })
        
        return {
            'statusCode': 200,
            'body': json.dumps({
                'branches': sorted(results, key=lambda x: x['distance'])
            })
        }
    except Exception as e:
        return {'statusCode': 500, 'body': json.dumps({'error': str(e)})}
      `),
      timeout: cdk.Duration.seconds(30),
    });

    const routeMatrixToolLambda = new lambda.Function(this, 'RouteMatrixToolLambda', {
      functionName: 'gym-route-matrix-tool',
      runtime: lambda.Runtime.PYTHON_3_10,
      handler: 'index.handler',
      code: lambda.Code.fromInline(`
import json
import boto3

location_client = boto3.client('location')

def handler(event, context):
    try:
        body = json.loads(event['body']) if 'body' in event else event
        user_coord = body['userCoordinate']  # [lat, lon]
        branch_coords = body['branchCoordinates']  # [[lat, lon], ...]
        
        # Use Amazon Location Service to calculate route matrix
        response = location_client.calculate_route_matrix(
            CalculatorName='${this.stackName}-route-calculator',
            DeparturePositions=[user_coord],
            DestinationPositions=branch_coords,
            TravelMode='Car',
            DistanceUnit='Kilometers',
            DurationUnit='Seconds'
        )
        
        results = []
        for i, route in enumerate(response['RouteMatrix'][0]):
            if 'DurationSeconds' in route and 'DistanceKilometers' in route:
                results.append({
                    'branchIndex': i,
                    'etaMinutes': round(route['DurationSeconds'] / 60, 1),
                    'distanceKm': round(route['DistanceKilometers'], 2)
                })
        
        return {
            'statusCode': 200,
            'body': json.dumps({'routes': results})
        }
    except Exception as e:
        return {'statusCode': 500, 'body': json.dumps({'error': str(e)})}
      `),
      timeout: cdk.Duration.seconds(30),
    });

    // Amazon Location Service
    const routeCalculator = new location.CfnRouteCalculator(this, 'RouteCalculator', {
      calculatorName: `${this.stackName}-route-calculator`,
      dataSource: 'Here',
    });

    // API Gateway
    const api = new apigateway.RestApi(this, 'GymAvailabilityApi', {
      restApiName: 'gym-availability-api',
      defaultCorsPreflightOptions: {
        allowOrigins: apigateway.Cors.ALL_ORIGINS,
        allowMethods: apigateway.Cors.ALL_METHODS,
        allowHeaders: ['Content-Type', 'X-Amz-Date', 'Authorization', 'X-Api-Key'],
      },
    });

    // API Routes
    const branchesResource = api.root.addResource('branches');
    branchesResource.addMethod('GET', new apigateway.LambdaIntegration(apiLambda));

    const toolsResource = api.root.addResource('tools');
    const availabilityResource = toolsResource.addResource('availability');
    availabilityResource.addMethod('POST', new apigateway.LambdaIntegration(availabilityToolLambda));
    
    const routeMatrixResource = toolsResource.addResource('route-matrix');
    routeMatrixResource.addMethod('POST', new apigateway.LambdaIntegration(routeMatrixToolLambda));

    // IoT Rule to trigger ingest Lambda
    const iotRule = new iot.CfnTopicRule(this, 'MachineStatusRule', {
      ruleName: 'GymMachineStatusRule',
      topicRulePayload: {
        sql: "SELECT * FROM 'org/+/machines/+/status'",
        actions: [
          {
            lambda: {
              functionArn: ingestLambda.functionArn,
            },
          },
        ],
      },
    });

    // Permissions
    machineStateTable.grantReadWriteData(ingestLambda);
    aggregatesTable.grantReadWriteData(ingestLambda);
    machineStateTable.grantReadData(apiLambda);
    machineStateTable.grantReadData(availabilityToolLambda);

    routeMatrixToolLambda.addToRolePolicy(new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['geo:CalculateRouteMatrix'],
      resources: [routeCalculator.attrArn],
    }));

    ingestLambda.addPermission('AllowIoTInvoke', {
      principal: new iam.ServicePrincipal('iot.amazonaws.com'),
      sourceArn: iotRule.attrArn,
    });

    // WebSocket API for real-time updates (placeholder)
    // TODO: Implement WebSocket API Gateway V2 for live machine status updates
    // - Connection management Lambda
    // - Broadcast Lambda for real-time notifications
    // - Connection store in DynamoDB

    // Bedrock Agent Configuration (placeholder)
    // TODO: Configure Bedrock Agent with Converse API
    // - Knowledge base for tool schemas
    // - Agent with tool-use capabilities
    // - Integration with availability and route-matrix tools
    const bedrockAgentRole = new iam.Role(this, 'BedrockAgentRole', {
      assumedBy: new iam.ServicePrincipal('bedrock.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonBedrockFullAccess')
      ],
      inlinePolicies: {
        'LambdaInvokePolicy': new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: ['lambda:InvokeFunction'],
              resources: [
                availabilityToolLambda.functionArn,
                routeMatrixToolLambda.functionArn
              ]
            })
          ]
        })
      }
    });

    // Add Bedrock permissions to tool Lambdas
    availabilityToolLambda.grantInvoke(new iam.ServicePrincipal('bedrock.amazonaws.com'));
    routeMatrixToolLambda.grantInvoke(new iam.ServicePrincipal('bedrock.amazonaws.com'));

    // Outputs
    new cdk.CfnOutput(this, 'ApiGatewayUrl', {
      value: api.url,
      description: 'API Gateway endpoint URL',
    });

    new cdk.CfnOutput(this, 'MachineStateTableName', {
      value: machineStateTable.tableName,
      description: 'DynamoDB table for machine states',
    });

    new cdk.CfnOutput(this, 'RouteCalculatorName', {
      value: routeCalculator.calculatorName,
      description: 'Amazon Location Route Calculator name',
    });
  }
}

File: gym_pulse/gym_pulse_stack.py
"""
GymPulse CDK Stack - Python Implementation
Real-time gym equipment availability with IoT integration and AI chatbot
"""
from aws_cdk import (
    Duration,
    Stack,
    CfnOutput,
    aws_dynamodb as dynamodb,
    aws_lambda as _lambda,
    aws_apigateway as apigateway,
    aws_apigatewayv2 as apigatewayv2,
    aws_iot as iot,
    aws_iam as iam,
    aws_location as location,
)
from constructs import Construct


class GymPulseStack(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # ========================================
        # DynamoDB Tables
        # ========================================
        
        # Current state table for real-time machine status
        current_state_table = dynamodb.Table(
            self, "CurrentStateTable",
            table_name="gym-pulse-current-state",
            partition_key=dynamodb.Attribute(
                name="machineId", 
                type=dynamodb.AttributeType.STRING
            ),
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            stream=dynamodb.StreamViewType.NEW_AND_OLD_IMAGES,
        )

        # Time-series events table for historical tracking
        events_table = dynamodb.Table(
            self, "EventsTable",
            table_name="gym-pulse-events",
            partition_key=dynamodb.Attribute(
                name="machineId", 
                type=dynamodb.AttributeType.STRING
            ),
            sort_key=dynamodb.Attribute(
                name="timestamp", 
                type=dynamodb.AttributeType.NUMBER
            ),
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            time_to_live_attribute="ttl",
        )

        # Aggregates table for analytics and heatmaps
        aggregates_table = dynamodb.Table(
            self, "AggregatesTable",
            table_name="gym-pulse-aggregates",
            partition_key=dynamodb.Attribute(
                name="gymId_category", 
                type=dynamodb.AttributeType.STRING
            ),  # format: "hk-central#legs"
            sort_key=dynamodb.Attribute(
                name="timestamp15min", 
                type=dynamodb.AttributeType.NUMBER
            ),
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            time_to_live_attribute="ttl",
        )

        # Alerts table for user notifications
        alerts_table = dynamodb.Table(
            self, "AlertsTable",
            table_name="gym-pulse-alerts",
            partition_key=dynamodb.Attribute(
                name="userId", 
                type=dynamodb.AttributeType.STRING
            ),
            sort_key=dynamodb.Attribute(
                name="machineId", 
                type=dynamodb.AttributeType.STRING
            ),
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
        )
        
        # Add GSI to alerts table
        alerts_table.add_global_secondary_index(
            index_name="machineId-index",
            partition_key=dynamodb.Attribute(
                name="machineId",
                type=dynamodb.AttributeType.STRING
            ),
        )

        # ========================================
        # IoT Core Infrastructure
        # ========================================
        
        # IoT device policy for MQTT publishing
        iot_policy = iot.CfnPolicy(
            self, "GymDevicePolicy",
            policy_name="GymDevicePolicy",
            policy_document={
                "Version": "2012-10-17",
                "Statement": [
                    {
                        "Effect": "Allow",
                        "Action": ["iot:Connect"],
                        "Resource": f"arn:aws:iot:{self.region}:{self.account}:client/gym-*",
                    },
                    {
                        "Effect": "Allow",
                        "Action": ["iot:Publish"],
                        "Resource": f"arn:aws:iot:{self.region}:{self.account}:topic/org/*/machines/*/status",
                    },
                    {
                        "Effect": "Allow",
                        "Action": ["iot:UpdateThingShadow"],
                        "Resource": f"arn:aws:iot:{self.region}:{self.account}:thing/*",
                    },
                ],
            },
        )

        # ========================================
        # Lambda Functions
        # ========================================
        
        # IoT message ingest and processing
        ingest_lambda = _lambda.Function(
            self, "IngestLambda",
            function_name="gym-pulse-iot-ingest",
            runtime=_lambda.Runtime.PYTHON_3_10,
            handler="handler.lambda_handler",
            code=_lambda.Code.from_asset("lambda/iot-ingest"),
            environment={
                "CURRENT_STATE_TABLE": current_state_table.table_name,
                "EVENTS_TABLE": events_table.table_name,
                "AGGREGATES_TABLE": aggregates_table.table_name,
                "ALERTS_TABLE": alerts_table.table_name,
            },
            timeout=Duration.seconds(30),
            memory_size=256,
        )

        # API handler for REST endpoints
        api_lambda = _lambda.Function(
            self, "ApiLambda",
            function_name="gym-pulse-api-handler",
            runtime=_lambda.Runtime.PYTHON_3_10,
            handler="handler.lambda_handler",
            code=_lambda.Code.from_asset("lambda/api-handlers"),
            environment={
                "CURRENT_STATE_TABLE": current_state_table.table_name,
                "EVENTS_TABLE": events_table.table_name,
                "AGGREGATES_TABLE": aggregates_table.table_name,
                "ALERTS_TABLE": alerts_table.table_name,
            },
            timeout=Duration.seconds(30),
            memory_size=512,
        )

        # WebSocket connection handler
        websocket_connect_lambda = _lambda.Function(
            self, "WebSocketConnectLambda",
            function_name="gym-pulse-websocket-connect",
            runtime=_lambda.Runtime.PYTHON_3_10,
            handler="connect.lambda_handler",
            code=_lambda.Code.from_asset("lambda/websocket-handlers"),
            timeout=Duration.seconds(30),
        )

        # WebSocket disconnect handler
        websocket_disconnect_lambda = _lambda.Function(
            self, "WebSocketDisconnectLambda",
            function_name="gym-pulse-websocket-disconnect",
            runtime=_lambda.Runtime.PYTHON_3_10,
            handler="disconnect.lambda_handler",
            code=_lambda.Code.from_asset("lambda/websocket-handlers"),
            timeout=Duration.seconds(30),
        )

        # Bedrock tool: getAvailabilityByCategory
        availability_tool_lambda = _lambda.Function(
            self, "AvailabilityToolLambda",
            function_name="gym-pulse-availability-tool",
            runtime=_lambda.Runtime.PYTHON_3_10,
            handler="availability.lambda_handler",
            code=_lambda.Code.from_asset("lambda/bedrock-tools"),
            environment={
                "CURRENT_STATE_TABLE": current_state_table.table_name,
                "AGGREGATES_TABLE": aggregates_table.table_name,
            },
            timeout=Duration.seconds(30),
        )

        # Bedrock tool: getRouteMatrix
        route_matrix_tool_lambda = _lambda.Function(
            self, "RouteMatrixToolLambda",
            function_name="gym-pulse-route-matrix-tool",
            runtime=_lambda.Runtime.PYTHON_3_10,
            handler="route_matrix.lambda_handler",
            code=_lambda.Code.from_asset("lambda/bedrock-tools"),
            environment={
                "ROUTE_CALCULATOR_NAME": "gym-pulse-route-calculator",
            },
            timeout=Duration.seconds(30),
        )

        # ========================================
        # Amazon Location Service
        # ========================================
        
        # Route calculator for ETA computation
        route_calculator = location.CfnRouteCalculator(
            self, "RouteCalculator",
            calculator_name="gym-pulse-route-calculator",
            data_source="Here",  # Using HERE as data provider
        )

        # ========================================
        # API Gateway Setup
        # ========================================
        
        # REST API
        api = apigateway.RestApi(
            self, "GymPulseApi",
            rest_api_name="gym-pulse-api",
            description="GymPulse Real-time Gym Availability API",
            default_cors_preflight_options=apigateway.CorsOptions(
                allow_origins=apigateway.Cors.ALL_ORIGINS,
                allow_methods=apigateway.Cors.ALL_METHODS,
                allow_headers=["Content-Type", "X-Amz-Date", "Authorization", "X-Api-Key", "X-Amz-Security-Token"],
            ),
        )

        # WebSocket API for real-time updates  
        websocket_api = apigatewayv2.WebSocketApi(
            self, "GymPulseWebSocketApi",
            route_selection_expression="$request.body.action",
        )

        # ========================================
        # API Routes
        # ========================================
        
        # REST API routes
        branches_resource = api.root.add_resource("branches")
        branches_resource.add_method("GET", apigateway.LambdaIntegration(api_lambda))
        
        branch_resource = branches_resource.add_resource("{id}")
        categories_resource = branch_resource.add_resource("categories")
        category_resource = categories_resource.add_resource("{category}")
        machines_resource = category_resource.add_resource("machines")
        machines_resource.add_method("GET", apigateway.LambdaIntegration(api_lambda))
        
        machines_root = api.root.add_resource("machines")
        machine_resource = machines_root.add_resource("{id}")
        history_resource = machine_resource.add_resource("history")
        history_resource.add_method("GET", apigateway.LambdaIntegration(api_lambda))
        
        alerts_resource = api.root.add_resource("alerts")
        alerts_resource.add_method("POST", apigateway.LambdaIntegration(api_lambda))

        # WebSocket routes (simplified for initial deployment)
        # TODO: Add WebSocket integrations in Phase 4
        websocket_stage = apigatewayv2.WebSocketStage(
            self, "WebSocketStage",
            web_socket_api=websocket_api,
            stage_name="prod",
            auto_deploy=True
        )

        # ========================================
        # IoT Rule for Lambda Trigger
        # ========================================
        
        iot_rule = iot.CfnTopicRule(
            self, "MachineStatusRule",
            rule_name="GymPulseMachineStatusRule",
            topic_rule_payload=iot.CfnTopicRule.TopicRulePayloadProperty(
                sql="SELECT *, topic() as topic FROM 'org/+/machines/+/status'",
                actions=[
                    iot.CfnTopicRule.ActionProperty(
                        lambda_=iot.CfnTopicRule.LambdaActionProperty(
                            function_arn=ingest_lambda.function_arn,
                        ),
                    ),
                ],
                error_action=iot.CfnTopicRule.ActionProperty(
                    republish=iot.CfnTopicRule.RepublishActionProperty(
                        role_arn="arn:aws:iam::123456789012:role/iot-error-role",
                        topic="error/gym-pulse"
                    )
                )
            ),
        )

        # ========================================
        # IAM Permissions
        # ========================================
        
        # DynamoDB permissions
        current_state_table.grant_read_write_data(ingest_lambda)
        events_table.grant_read_write_data(ingest_lambda)
        aggregates_table.grant_read_write_data(ingest_lambda)
        alerts_table.grant_read_write_data(ingest_lambda)
        
        current_state_table.grant_read_data(api_lambda)
        events_table.grant_read_data(api_lambda)
        aggregates_table.grant_read_data(api_lambda)
        alerts_table.grant_read_write_data(api_lambda)
        
        current_state_table.grant_read_data(availability_tool_lambda)
        aggregates_table.grant_read_data(availability_tool_lambda)

        # IoT Core permissions
        ingest_lambda.add_permission(
            "AllowIoTInvoke",
            principal=iam.ServicePrincipal("iot.amazonaws.com"),
            source_arn=iot_rule.attr_arn,
        )

        # Location Service permissions
        route_matrix_tool_lambda.add_to_role_policy(
            iam.PolicyStatement(
                effect=iam.Effect.ALLOW,
                actions=["geo:CalculateRouteMatrix"],
                resources=[route_calculator.attr_arn]
            )
        )

        # Bedrock agent role for tool invocation
        bedrock_agent_role = iam.Role(
            self, "BedrockAgentRole",
            assumed_by=iam.ServicePrincipal("bedrock.amazonaws.com"),
            inline_policies={
                "LambdaInvokePolicy": iam.PolicyDocument(
                    statements=[
                        iam.PolicyStatement(
                            effect=iam.Effect.ALLOW,
                            actions=["lambda:InvokeFunction"],
                            resources=[
                                availability_tool_lambda.function_arn,
                                route_matrix_tool_lambda.function_arn
                            ]
                        )
                    ]
                )
            }
        )

        # Grant Bedrock permission to invoke tool Lambdas
        availability_tool_lambda.grant_invoke(iam.ServicePrincipal("bedrock.amazonaws.com"))
        route_matrix_tool_lambda.grant_invoke(iam.ServicePrincipal("bedrock.amazonaws.com"))

        # ========================================
        # CloudFormation Outputs
        # ========================================
        
        CfnOutput(
            self, "ApiGatewayUrl",
            value=api.url,
            description="REST API Gateway endpoint URL",
        )

        CfnOutput(
            self, "WebSocketApiUrl", 
            value=f"wss://{websocket_api.api_id}.execute-api.{self.region}.amazonaws.com/{websocket_stage.stage_name}",
            description="WebSocket API endpoint URL",
        )

        CfnOutput(
            self, "CurrentStateTableName",
            value=current_state_table.table_name,
            description="DynamoDB table for current machine states",
        )

        CfnOutput(
            self, "EventsTableName",
            value=events_table.table_name,
            description="DynamoDB table for time-series events",
        )

        CfnOutput(
            self, "AvailabilityToolArn",
            value=availability_tool_lambda.function_arn,
            description="Availability tool Lambda ARN for Bedrock agent",
        )

        CfnOutput(
            self, "RouteMatrixToolArn",
            value=route_matrix_tool_lambda.function_arn,
            description="Route matrix tool Lambda ARN for Bedrock agent",
        )

        CfnOutput(
            self, "RouteCalculatorName",
            value=route_calculator.calculator_name,
            description="Amazon Location Route Calculator name",
        )
